### ISSUE TITLE
ReturnMissingNullable doesn't handle simple null flows correctly

### ISSUE THREAD

**Issue by vovkab** [google/error-prone#907]
### What version of Error Prone are you using?
2.2.0

### Gradle configuration:

``` gradle
classpath 'net.ltgt.gradle:gradle-errorprone-plugin:0.0.13'
errorprone 'com.google.errorprone:error_prone_core:2.2.0'

tasks.withType(JavaCompile) {
    options.compilerArgs += [
            '-Xep:ParameterNotNullable:ERROR',
            '-Xep:ReturnMissingNullable:ERROR',
    ]
}
```

### Does this issue reproduce with the latest release?
Yes

### What did you do?

```java
@Nullable
private String mText;

@NonNull
public String getText() {
    return mText == null ? "" : mText;
}
```

### What did you expect to see?
Should not show any nullable errors.

### What did you see instead?
    Error:(23, 9) error: [ReturnMissingNullable] May return null
    (see http://errorprone.info/bugpattern/ReturnMissingNullable)
    Did you mean '@Nullable @NonNull'?

**Comment by nglorioso** [comment]
Hello! This pattern is actually not thread-safe, since `mText` can be reassigned to null in a separate thread between the comparison to null and the ternary statement.

A pattern which is provably safe:

```
@Nullable
private String mText;

@NonNull
public String getText() {
    String currText = mText;
    return currText == null ? "" : currText;
}
```

Or, if you can make `mText` final, that should also work.

**Comment by vovkab** [comment]
@nglorioso Thanks! 
I thought I checked this use case too, maybe I confused with other errors that was shown.
Making it `final` or copying locally, like you suggested, works!

I'm adding this checks for android code and most of this code runs on main thread. Do you know if there is anyway to suppress thread check or maybe somehow let is know that this is the same thread?

**Comment by eaftan** [comment]
You can use `@SuppressWarnings("ReturnMissingNullable")` on that method to suppress.

**Comment by Maaartinus** [comment]
@eaftan But that's too generous, suppressing also single-threaded bugs like `return mText == null ? anotherNullableText : mText`. Some weaker suppression like `@SuppressWarnings("ReturnMissingNullableWhenMultithreaded")` would be handy.

Maybe also a message like *May return null due to concurrent modifications*, which would make such a case perfectly clear.

**Comment by vovkab** [comment]
Found another use case where it fails:

```java
@Nullable
private String mText;

@NonNull
public String getText() {
    final String text = mText;
    return isEmpty(text) ? "" : text;
}

public boolean isEmpty(@Nullable String text) {
    return text == null || text.isEmpty();
}
```

Fails with:
```
Error:(25, 9) error: [ReturnMissingNullable] May return null
(see http://errorprone.info/bugpattern/ReturnMissingNullable)
Did you mean '@Nullable @NonNull'?
```


And another one, that compiles without errors or warnings:
```java
void myNonNullMethod(@NonNull String param) {}
myNonNullMethod(null);
```

p.s. I know JetBrains provides `Contract` annotation for something like this, to help determine nullability, i.e. `@Contract("null -> true")`. Is there any alternative in ErrorProne, or maybe there is a plans to support `Contract`?

**Comment by kevin1e100** [comment]
Thanks for the suggestions.  For context, this "check" is intended as a refactoring help, to mechanically propagate @Nullable annotations around your code.  It wasn't conceived as a bug checker per se (note that it looks for missing annotations, not possible null dereferences) though we know that people are using it as a bug checker and are evaluating how to support that better (and regardless, false positives are always annoying).  That's also the reason why this "check" is off by default.  Meanwhile, there are as you point out other tools that are more targeted at static checking around null pointer dereferences that may just be more appropriate to use in your case.

A couple clarifications:
- the check does support common Guava methods that test for null, such as com.google.common.base.Strings.isNullOrEmpty, which is equivalent to the example above.  The check just doesn't support user-written null test methods.
- the pattern "x != null ? x : default" should be handled correctly when x is a local; it's just that fields are much harder to track correctly in the general case, so the check gives up there for simplicity (as opposed to being particularly conservative for concurrency issues).  If we tracked fields we could handle that case better, especially as you say when assuming the absence of race conditions.

**Comment by yborovikov** [comment]
Without the `final` modifier raising the `ReturnMissingNullable` is proper.
For `final` fields (should be treated like local variables), however, it's a false-positive:

```java
class Example {
    final @Nullable String name;
    Example(@Nullable String name) { this.name = name; }
    public @NotNull String toString() { return name == null ? "" : name; }
}
```
error_prone_core `2.2.0`, `2.3.0`, `2.3.1` - all yield a `ReturnMissingNullable` for the method.

**Comment by cushon** [comment]
c8c034d99b48df0034aad64ea777b480f6c3cbf9
